lib/connection_controller.go

```diff
@@ -0,0 +1,330 @@
+func (cc *ConnectionController) _handleNewConnectionMessage(origin *Peer, desoMsg DeSoMessage) {
+	if desoMsg.GetMsgType() != MsgTypeNewConnection {
+		return
+	}
+
+	var msg *MsgDeSoNewConnection
+	var ok bool
+	if msg, ok = desoMsg.(*MsgDeSoNewConnection); !ok {
+		return
+	}
+
+	var remoteNode *RemoteNode
+	var err error
+	switch msg.Connection.GetConnectionType() {
+	case ConnectionTypeInbound:
+		remoteNode, err = cc.processInboundConnection(msg.Connection)
+		if err != nil {
+			glog.Errorf("ConnectionController.handleNewConnectionMessage: Problem handling inbound connection: %v", err)
+			msg.Connection.Close()
+			return
+		}
+	case ConnectionTypeOutbound:
+		remoteNode, err = cc.processOutboundConnection(msg.Connection)
+		if err != nil {
+			glog.Errorf("ConnectionController.handleNewConnectionMessage: Problem handling outbound connection: %v", err)
+			cc.cleanupFailedOutboundConnection(msg.Connection)
+			return
+		}
+	}
+
+	// If we made it here, we have a valid remote node. We will now initiate the handshake.
+	cc.handshake.InitiateHandshake(remoteNode)
+}
+
+func (cc *ConnectionController) cleanupFailedOutboundConnection(connection Connection) {
+	oc, ok := connection.(*outboundConnection)
+	if !ok {
+		return
+	}
+
+	id := NewRemoteNodeId(oc.attemptId)
+	rn := cc.rnManager.GetRemoteNodeById(id)
+	if rn != nil {
+		cc.rnManager.Disconnect(rn)
+	}
+	cc.cmgr.RemoveAttemptedOutboundAddrs(oc.address)
+}
+
+// ###########################
+// ## Connections
+// ###########################
+
+func (cc *ConnectionController) CreateValidatorConnection(ipStr string, publicKey *bls.PublicKey) (_err error) {
+	netAddr, err := cc.ConvertIPStringToNetAddress(ipStr)
+	if err != nil {
+		return err
+	}
+	return cc.rnManager.CreateValidatorConnection(netAddr, publicKey)
+}
+
+func (cc *ConnectionController) CreateNonValidatorPersistentOutboundConnection(ipStr string) (_err error) {
+	netAddr, err := cc.ConvertIPStringToNetAddress(ipStr)
+	if err != nil {
+		return err
+	}
+	return cc.rnManager.CreateNonValidatorPersistentOutboundConnection(netAddr)
+}
+
+func (cc *ConnectionController) CreateNonValidatorOutboundConnection(ipStr string) (_err error) {
+	netAddr, err := cc.ConvertIPStringToNetAddress(ipStr)
+	if err != nil {
+		return err
+	}
+	return cc.rnManager.CreateNonValidatorOutboundConnection(netAddr)
+}
+
+func (cc *ConnectionController) SetTargetOutboundPeers(numPeers uint32) {
+	cc.targetOutboundPeers = numPeers
+}
+
+func (cc *ConnectionController) enoughNonValidatorInboundConnections() bool {
+	return uint32(cc.rnManager.GetNonValidatorInboundIndex().Count()) >= cc.maxInboundPeers
+}
+
+func (cc *ConnectionController) enoughNonValidatorOutboundConnections() bool {
+	return uint32(cc.rnManager.GetNonValidatorOutboundIndex().Count()) >= cc.targetOutboundPeers
+}
+
+// processInboundConnection is called when a new inbound connection is established. It is responsible for validating
+// the connection and creating a RemoteNode from it.
```

Comment:
What does it mean for a new inbound connection to have been established in this case? Does it mean that the network connection has been dialed up? Has the handshake process started? Please clarify at what stage of the connection lifecycle this is called

Reply:
Hmm, I think technically "established" is correct in this context. We're processing a newly established connection, albeit at this point it is completely anonymous to us. Anyway, I added some more words of explanation.